os/binfmt/binfmt_loadlibrary.c

```diff
@@ -0,0 +1,134 @@
+#include <tinyara/binary_manager.h>
+
+#include "binfmt.h"
+#include "binary_manager/binary_manager.h"
+#include <tinyara/mm/mm.h>
+
+#if defined(CONFIG_BINFMT_ENABLE) && defined(CONFIG_SUPPORT_COMMON_BINARY)
+
+struct binary_s *g_lib_binp;
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: load_library
+ *
+ * Description:
+ *   This is a convenience function that wraps load_module and other steps into
+ *   one call and is used to load a library and export its symbols.
+ *
+ * Input Parameters:
+ *   filename  - The path to the program to be executed. If
+ *               CONFIG_LIB_ENVPATH is defined in the configuration, then
+ *               this may be a relative path from the current working
+ *               directory. Otherwise, path must be the absolute path to the
+ *               program.
+ *
+ * Returned Value:
+ *   This is an end-user function, so it follows the normal convention:
+ *   It returns the PID of the exec'ed module.  On failure, it returns
+ *   -1 (ERROR) and sets errno appropriately.
+ *
+ ****************************************************************************/
+int load_library(FAR const char *filename)
+{
+	FAR struct binary_s *bin = NULL;
+	int errcode;
+	int ret;
+
+#ifdef CONFIG_OPTIMIZE_APP_RELOAD_TIME
+	bin = g_lib_binp;
+	/* If we find a non-null value for bin, it means that
+	 * we are in a reload scenario.
+	 */
+	if (bin) {
+		if (!bin->data_backup) {
+			errcode = -EINVAL;
+			berr("ERROR: Failed to find copy of data section from previous load\n");
+			goto errout_with_bin;
+		}
+
+		memcpy(bin->datastart, bin->data_backup, bin->datasize);
+		memset(bin->bssstart, 0, bin->bsssize);
+	} else {
+#endif
+
+		/* Allocate the load information */
+
+		bin = (FAR struct binary_s *)kmm_zalloc(sizeof(struct binary_s));
+		if (!bin) {
+			berr("ERROR: Failed to allocate binary_s\n");
+			errcode = -ENOMEM;
+			goto errout_with_bin;
+		}
+
+		/* Initialize the binary structure */
+		bin->filename = filename;
+		bin->islibrary = true;
+		g_lib_binp = bin;
+
+		/* Load the module into memory */
+		ret = load_module(bin);
+		if (ret < 0) {
+			errcode = -ret;
+			berr("ERROR: Failed to load program '%s': %d\n", filename, errcode);
+			goto errout_with_bin;
+		}
+
+#ifdef CONFIG_OPTIMIZE_APP_RELOAD_TIME
+		if (!bin->data_backup) {
+			errcode = -EINVAL;
+			berr("ERROR: data section backup address not initialized\n");
+			goto errout_with_bin;
+		}
+
+		memcpy(bin->data_backup, bin->datastart, bin->datasize);
+	}
+#endif
+
+	return OK;
+
+errout_with_unload:
+	(void)unload_module(bin);
+errout_with_bin:
+	kmm_free(bin);
+	g_lib_binp = NULL;
+errout:
+	set_errno(errcode);
+	return ERROR;
+
+}
+#endif							/* if defined(CONFIG_BINFMT_ENABLE) && defined(CONFIG_SUPPORT_COMMON_BINARY) */
```

Comment:
This code is similar to the load_binary API in "binfmt/binfmt_loadbinary.c"
Is it necessary to keep it separate or can we merge the two with a binary/library parameter to a common API such as:
load_module(int binary_idx, FAR const char *filename, load_attr_t *load_attr)
where load_attr could be > binary/library

Reply:
You are right. Initially i expected many changes in library loading. So, I had made new API. 
But, i feel it can be merged with loadbinary api. 

Done.